libraries/llvm/rts.ll

```diff
@@ -366,30 +366,63 @@ define private %Reference @newReference(%Stack %stack) alwaysinline {
 ; Stack management
 
 define private %StackPointer @stackAllocate(%Stack %stack, i64 %n) {
-    %stackStackPointer = getelementptr %StackValue, %Stack %stack, i64 0, i32 1, i32 0
-    %stackPointer = load %StackPointer, ptr %stackStackPointer
+    %stackPointer_pointer = getelementptr %StackValue, %Stack %stack, i64 0, i32 1, i32 0
 
-    %stackPointer_2 = getelementptr i8, %StackPointer %stackPointer, i64 %n
-    store %StackPointer %stackPointer_2, ptr %stackStackPointer
-    ret %StackPointer %stackPointer
+    %stackMemory = getelementptr %StackValue, %Stack %stack, i64 0, i32 1
+    %memory = load %Memory, ptr %stackMemory
+
+    %current = extractvalue %Memory %memory, 0
+    %limit = extractvalue %Memory %memory, 2
+
+    %nextStackPointer = getelementptr i8, %StackPointer %current, i64 %n
+    %cmp = icmp ule %StackPointer %nextStackPointer, %limit
+    br i1 %cmp, label %continue, label %realloc
+
+continue:
+    store %StackPointer %nextStackPointer, ptr %stackPointer_pointer
+    ret %StackPointer %current
+
+realloc:
+    %base = extractvalue %Memory %memory, 1
+
+    %intStackPointer = ptrtoint %StackPointer %current to i64
+    %intBase = ptrtoint %Base %base to i64
+
+    %size = sub i64 %intStackPointer, %intBase
+    %double = mul i64 %size, 2
+    %newSize = add i64 %double, %n ; TODO: should we be smarter here?
```

Comment:
As someone who doesn't understand the LLVM runtime, why do we need to add the `%n` here?

Reply:
It's not strictly related to the runtime, but rather the growing logic. Since the bytes to be allocated (`n`) could be larger than the double of the current stack size, I first double the size and then add `n` to it. We could (should?) of course do something smarter here.

Reply:
Intuitively, I'd expect something like "if we know we have to resize, then `new_size <- nextStrictlyBiggestPowerOfTwo(current_size, n)" where the mysterious function is https://llvm.org/doxygen/namespacellvm.html#afb65eef479f0473d0fe1666b80155237 or `clz` (get the highest bit, choose a number one bigger in binary), but I have to think about it.

Reply:
Yes, that also sounds good. My thought behind the strategy was that, intuitively, large stack allocations are quite rare. So if I allocate 1GB on a 10MB stack, I'd prefer the next stack size to be 20MB + 1GB rather than 2GB.

Reply:
Might be worth looking at 

https://github.com/golang/go/blob/24cb743d1faa6c8f612faa3c17ac9de5cc385832/src/runtime/stack.go#L336

and it's callsites.

Reply:
Very interesting, although I agree that maybe you're overthinking this a bit :octocat:

Regarding the profile of the allocations: We (currently?) only use `@stackAllocate` to make space for pushes of pos/neg, builtin types and other pointers, never for arbitrarily large data allocations. (Maybe @phischu can confirm this?)

As far as I understand, large (de-)allocations only happen when some function has a lot of arguments that need to be stored on the stack (~8-16 byte each). In `nqueens`, for example, this leads to stack (de-)allocations of 108 bytes.

To make this more clear: Allocation sizes/amount in all generated files in `effekt.llvmtests`:

|bytes allocated|how often|
|-----|-----|
|8|318|
|16|305|
|24|1856|
|32|101|
|40|57|
|48|22|
|56|24|
|64|4|
|72|1|
|80|1|

So I don't think we really need to be creative with the growing logic, since `%n` will almost never be larger than the doubled current stack size, especially if we settle on an initial stack size like 1024. Minimal solutions like adding the current size (as I did), or by using `NextPowerOf2` are probably more than enough.

Reply:
@phischu asked about the stack allocation profile at runtime. Here's the data:

For the entire test suite (`test`):

| bytes allocated | how often |
|-----------------|-----------|
| 8               | 2918774   |
| 16              | 815964    |
| 24              | 5386483   |
| 32              | 841972    |
| 40              | 169619    |
| 48              | 6444      |
| 56              | 79        |
| 64              | 2085      |
| 72              | 126       |
| 80              | 707       |
| 96              | 8         |
| 112             | 1         |

Only the LLVM tests (`testOnly effekt.LLVMTest`)

| bytes allocated | how often |
|-----------------|-----------|
| 8               | 2918721   |
| 16              | 815664    |
| 24              | 5385418   |
| 32              | 841852    |
| 40              | 169604    |
| 48              | 6435      |
| 56              | 79        |
| 64              | 2080      |
| 72              | 126       |
| 80              | 702       |